runtime/runtime/src/ext.rs

```diff
@@ -254,6 +276,40 @@ impl<'a> External for RuntimeExt<'a> {
         Ok(())
     }
 
+    #[cfg(feature = "protocol_feature_function_call_ratio")]
+    fn append_action_function_call_ratio(
+        &mut self,
+        receipt_index: u64,
+        method_name: Vec<u8>,
+        args: Vec<u8>,
+        attached_deposit: u128,
+        prepaid_gas: u64,
+        gas_ratio: u64,
+    ) -> ExtResult<()> {
+        let action_index = self.append_action(
+            receipt_index,
+            Action::FunctionCall(FunctionCallAction {
+                method_name: String::from_utf8(method_name)
+                    .map_err(|_| HostError::InvalidMethodName)?,
+                args,
+                gas: prepaid_gas,
+                deposit: attached_deposit,
+            }),
+        );
+
+        if gas_ratio > 0 {
+            self.distribute_leftover_gas_to.push(GasRatioMetadata {
+                receipt_index: receipt_index as usize,
+                action_index,
+                gas_ratio,
+            });
+            self.gas_ratio_sum =
+                self.gas_ratio_sum.checked_add(gas_ratio).ok_or(HostError::IntegerOverflow)?;
```

Comment:
This error condition doesn't feel great to me. This is a code path which could be triggered by the user, it's not just dead code, so it adds to the number of code paths which we need to exercise. "You can't use `u64::MAX - 1` as a ratio" is also an error condition the user needs to worry about. Let's implement the logic in a way which can deal with arbitrary ratios. Using `u128` to sum the ratios should work I think.

Reply:
Seems like a bit of a foot-gun because the gas maximum is u64::MAX, so having the ratio larger doesn't do anything and goes into the fallback case, which could silently cause unexpected behaviour. Should never realistically happen so maybe not worth handling this case.